### ISSUE TITLE
Use 308 instead of 301 as default https redirect code

### ISSUE THREAD

**Issue by johnlanni** [envoyproxy/envoy#22973]
*Title*: *Use 308 instead of 301 as default https redirect code*

*Description*:

When `require_tls` is set to `ALL` in [VirutalHost](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html?highlight=virtualhost#config-route-v3-virtualhost), http requests will be redirected to https with a 301 code. I think using 308 is a better choice.

[RFC 7238](https://tools.ietf.org/html/rfc7238) was created to define the 308 (Permanent Redirect) status code that is similar to 301 (Moved Permanently) but it keeps the payload in the redirect. This is important if we send a redirect in methods like POST.

**Comment by mattklein123** [comment]
cc @RyanTheOptimist @alyssawilk

**Comment by alyssawilk** [comment]
Section 6 of [RFC7231] requires recipients to treat unknown 3xx
   status codes the same way as status code 300 Multiple Choices
   ([RFC7231], Section 6.4.1).  Thus, servers will not be able to rely
   on automatic redirection happening similar to status codes 301, 302,
   or 307.

Happy to get a second opinion but I don't think switching the default makes sense given it looks like it could result in observable behavioral changes for internet facing clients.  I could see making it a configuration option but offhand I'd be inclined to suggest folks who want a 308 use the local reply rewrite options to return the code they prefer.

**Comment by RyanTheOptimist** [comment]
I agree with @alyssawilk. This does not seem like a change that would be safe to make unconditionally. An option, or every better, local reply rewrite seems like a better solution to me.

**Comment by johnlanni** [comment]
I prefer to add a new configuration option to set the status code, since setting a local reply is intrusive to the routing configuration, I would like to disable or enable this redirect behavior globally.

> Section 6 of [RFC7231] requires recipients to treat unknown 3xx status codes the same way as status code 300 Multiple Choices ([RFC7231], Section 6.4.1). Thus, servers will not be able to rely on automatic redirection happening similar to status codes 301, 302, or 307.
> 
> Happy to get a second opinion but I don't think switching the default makes sense given it looks like it could result in observable behavioral changes for internet facing clients. I could see making it a configuration option but offhand I'd be inclined to suggest folks who want a 308 use the local reply rewrite options to return the code they prefer.

**Comment by alyssawilk** [comment]
I believe the HCM supports local reply configuration as well, so you'd have to present a compelling reason to add custom code for this.

**Comment by johnlanni** [comment]
> Contributor



> I believe the HCM supports local reply configuration as well, so you'd have to present a compelling reason to add custom code for this.

This redirect behavior also requires setting the `Location` response header, which seems to be an intrusion into the routing configuration
